---
title: ''
author: "Jonathan Flowers"
date: "2023-09-21"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Assignment 1

### Task 1: Create a multiple sequence alignment

Q1.1. Upload your MSA with your homework assignment (4 points)

```{r}
library(phangorn)

seq.DNAbin <- read.dna("COX5A_aligned.fa", 
                       format = "fasta")

library(seqinr)
# Replace "path_to_your_fasta_file.fa" with the actual path to your FASTA file
fasta_data <- read.fasta(file = "path_to_your_fasta_file.fa")




```

Visualize the MSA and select the correct answer.

Q1.2 How many insertion/deletions are in your alignment (3 points)?  
(a) 1  
(b) 2  
(c) 6  
(d) 10  

In this case, the alignment is clean and there is no ambiguity in homology (few gaps, high sequence identity across the alignment). Therefore there is no need to run GBlocks or comparable tool to remove parts of the MSA with questionable homology.

### Task 2: Build a distance-based phylogeny

You are now ready to build a phylogenetic tree from the MSA using a distance method. Trees constructed in this fashion are similar to hierarchical clustering and may even use the same algorithms and non-evolutionary applications. However, distance-based phylogeny reconstruction typically uses an evolutionary rate model to adjust for known biases as we shall see.



Before running this block of code, copy the aligned fasta to same directory as this .Rmd file and update the `aligned_fasta` variable with your MSA filename. Then execute this code block to read the alignment.

```{r}
library(phangorn)
aligned_fasta <- 'COX5A_aligned.fa' # <- your MSA file here between single quotes
seq.DNAbin <- read.dna(aligned_fasta, 
                       format = "fasta")
seq.phyDat <- as.phyDat(seq.DNAbin)
```

Building a distance tree requires a distance matrix with evolutionary distance estimates between all pairs of taxa in the analysis. In distance-based tree reconstruction, it is often necessary to apply some form of correction to the distance estimate to account for the possiblity of 'multiple hits' (see lecture material), rate differences between transition and tranversion mutatons and other factors. Here, we will use the Jukes-Cantor (JC69) correction.

Create a distance matrix.

```{r}
seq.dist <- dist.ml(seq.phyDat,model="JC69")
seq.dist
```
Now lets see how much the JC69 distances compare to p-distances. P-distances are the standard hamming distance, or, the proportion of sites that differ between a pair of sequences.

```{r}
dist.hamming(seq.phyDat, ratio = TRUE, exclude = "none")
```
Q2.1 Review the two distance matrices. There are two pairs of sequences that are identical. Which are they? (2 points)  

Q2.2 Find both the p-distance and J69 distance between the human sequence Homo_sapiens_isolate_Hsap76 and Eulemur_fulvus. What are the uncorrected and JC69 distances, respectively? (2 points)  

Q2.3 Which of the following best explains why the JC69 distance is greater than the p-distance? (3 points)
(a) the JC69 distance includes indels, p-distances do not  
(b) the JC69 distance corrects for multiple hits, p-distances do not  
(c) the JC69 distance adjusts for alignment errors, the p-distance does not  
(d) all of the above  


Q2.4. **You will be awarded 1 point for simply executing the code and showing the tree in your report for all trees in this assignment.** You can build your first tree here by running the Neighbor-joining algorithm on the JC69 distance matrix (1 point)

```{r}
seq.NJ.phylo <- NJ(seq.dist)
seq.NJ.phylo$edge.length[seq.NJ.phylo$edge.length < 0] <- 0 # sometimes NJ algorithm have negative branch lengths, which are artifacts of the algorithm and can be safely set to zero

# Now plot using plot.phylo function
plot.phylo(seq.NJ.phylo,type="fan",main="Neighbor-Joining (unrooted)",cex = 0.5)
```



Phylogenetic trees are not generally rooted by default (although there are exceptions). Lets check if  the NJ tree above is rooted.
```{r}
ape::is.rooted(seq.NJ.phylo)
```

Q2.5 Now create a 'bootstrap tree' with bootstrap support values on each branch using the `bootstrap.phyDat` function. Note that we specify argument `bs=1000` to generate 1000 pseudoreplicate MSAs and trees by bootstrapping over columns in the MSA (1 point)

```{r}
seq.NJ.multiphylo <- bootstrap.phyDat(seq.phyDat, bs = 1000, function(x) NJ(dist.ml(seq.phyDat,model="JC69")))
plotBS(seq.NJ.phylo,seq.NJ.multiphylo,type = "fan",main="Neihbor-Joining Bootstrap Tree (unrooted)",cex = 0.5)
```

Q2.6. Bootstrapping in molecular phylogeny reconstruction involves sampling columns in the MSA with replacement. Which of the following statements is true about this bootstrap procedure and output tree. Select all true statements. (3 points)  

(a) Each 'pseudoreplicate' MSA has the same number of characters (i.e., columns) as the original MSA  
(b) No column from the original MSA can appear more than once in a pseudoreplicate MSA  
(c) A distance matrix is created for each pseudoreplicate MSA followed by tree construction for that   pseudoreplicate
(d) The bootstrap value on each branch of the bootstrap tree represents the percent of pseudoreplicate trees in which that branch was observed.  

Q2.7 Some distance-based algorithms are 'ultrametric' and assume equal rates of evolution (a 'molecular clock'). Such trees can be recognized because branch lengths of taxa that share a most recent common ancestor (MRCA) will have the same branch lengths leading back to the shared ancestor.  

In non-ultrametric trees, taxa that share an MRCA **may** have different branch lengths, or not. Since such taxa share a common ancestor at the same time in the past, any differences in branch lengths leading to the common ancestor reflects differences in rates of evolution (differences in the rate of accumulation of mutations) in the two species. 

Lemurs (represented by Eulemur fulvus) are the outgroup to monkeys, gibbons and great apes included in this study. Use this information to answer the following questions

Q2.7a Given this outgroup, on what branch of the tree should the root edge appear (2 points)?
(a) on the branch leading to the two human sequences
(b) on the branch leading to orangutans (Pongo)
(c) on the branch leading to old world monkeys (Colobus and Papio)
(d) on the branch leading to Eulemur fulvus

Q2.7b. List a pair of terminal taxa that form a clade (a monophyletic group), as indicated by a bootstrap value of 90 or higher **and** clearly have different rates of evolution (2 points).  

Q2.7c. Which of those two taxa has a faster rate of evolution on the terminal branch? (2 points)

Now we will try rooting the tree on the Lemur sequence (Eulemur fulvus). Lemurs are known outgroups to all species of old world monkeys, gibbons, and great apes which constitute the rest of the taxa in this analysis.
```{r}
library(ggtree)
library(treeio)
seq.NJ_rooted.treedata <- seq.NJ.phylo %>% 
  ape::root(outgroup = "Eulemur_fulvus", edgelabel = TRUE,resolve.root=T) %>%
  as_tibble() %>%
  as.treedata
```

Lets confirm tree is rooted.
```{r}
ape::is.rooted(seq.NJ_rooted.treedata)
```

Q2.8 Now lets plot the rooted original tree using ggtree package. Note that you will compare this output tree to the Maximum Likelihood tree in Task 3 (1 point)

```{r}
ggtree(seq.NJ_rooted.treedata,layout="rectangular") + 
  geom_tiplab() + 
  geom_rootpoint(size=2,color="red",shape=2)
```

The ape::root function joins the outgroup at the node "below" the outgroup as indicated here with a red triangle, but does not draw a root edge. Ideally, a root edge (branch) would be drawn on the branch leading to Eulemur fulvus creating a bifurcation between Eulemur and all members of the "ingroup". Nevertheless, using Eulemur fulvus as an outgroup, Papio anubis and Colobus guereza (old world monkeys) form a sister group Gibbons (Symphalangus, Nomascus), Orangutans (Pongo), and the great apes as expected.


### Task 3: Maximum Likelihood (ML) analysis of COX5A.

Once you have an appropriately curated MSA, the first step in ML-based phylogeny reconstruction is typically to identify a model of sequence evolution that best fits the data. **Some methods like the popular RaxML software just assume a General Time Reversible model is best, which may or may not be true**, but for large genome-scale datasets is sometimes assumed for computational ease. For smaller datasets, the best fit model should be determined with Model Test, which calculates the Akaike Information Criteria (AIC) and the Bayesian Information Criterion for a large number of substitution models. These criteria can then be used to select the best model of sequence evolution to explain the patterns of mutations in the MSA. 

Note, 'substitution' and 'mutation' are sometimes used loosely as equivalent in the phylogenetics vernacular, although we will see later in the course that substitution and mutation are different processes in molecular evolution. 

Run ModelTest here as implemented in the phangorn package.

```{r}
# Model selection
# note: mt is a "modelTest" data.frame with LogLik, AIC, BIC, etc for each model
mt <- modelTest(seq.phyDat)
```

Now review the output in sorted order according to AIC.

```{r}
mt %>%
  as_tibble %>%
  dplyr::arrange(AIC)
```

How do we identify model with best fit? The Akaike Information Criterion (AIC) and Bayesian Information Criterion (BIC) are approaches to identify the best model. Review the AIC and BIC wikipedia pages here <https://en.wikipedia.org/wiki/Akaike_information_criterion> and here <https://en.wikipedia.org/wiki/Bayesian_information_criterion>. 

Q3.1 Which model is the best fit according to AIC? (3 points)

(a) GTR+G(4)+I
(b) F81
(c) K80+I
(d) TPM3+I


Now we can run the ML analysis to determine the tree topology and branch lengths that maximize the probability of observing the sequences in the MSA.



```{r}
# note: here we provide a modelTest object instead of alignment
pml_bb.modelTest <- pml_bb(mt, control = pml.control(trace = 0), rearrangement = "stochastic")  # pml_bb.modelTest is a pml object
```

Q3.2 Review the pml_bb documentation "Details" (at your console, enter `?pml_bb`). What model selection method does this function use to select the best-fitting model of sequence evolution from the modelTest result object? (2 points)
(a) AIC
(b) BIC
(c) Maximum Likelihood
(c) none of the above

Once again, check if the tree is rooted
```{r}
# note: the tree is found "tree" slot of pml object and is an object of class phylo
pml_bb.modelTest$tree %>%
  ape::is.rooted()
```

Q3.3 Lets root using same approach as with NJ distance tree and plot the ML tree (1 point)  

```{r}
# note: we can plot a pml object with base R plot function or convert to treedata object and plot with ggtree
pml_bb.modelTest$tree %>% 
  ape::root(outgroup = "Eulemur_fulvus", edgelabel = TRUE,resolve.root=T) %>%
  as_tibble() %>%
  as.treedata() %>%
  ggtree(layout="rectangular") + 
    geom_tiplab() + 
    geom_rootpoint(size=2,color="red",shape=2)
```


Now compare this tree, to the NJ tree at the end of Task 2. If we take at face value the branching pattern in each analysis, which phylogeny correctly identifies the great apes as a clade? Which one does not? Note that Pongo (Orangutans) are great apes, but gibbons (Symphalangus and Nomascus) are not. (2 points)  


Optimality criterion methods like ML search the "tree universe" (either evaluating all of the tree topologies in an exhaustive search or a subset of possible tree topologies using a tree-searching algorithm) to identify the tree that maximizes the optimality score, in this case, the Maximum Likelihood.  

Ideally, we would conduct an exhaustive search of all trees in the tree universe to determine their likelihood, thereby ensuring that we infer the ML tree and do not end up with a suboptimal tree (which can occur if the tree-searching algorithm gets stuck in a local optimum). Your instructor couldn't find an option to force phangorn to perform an exhaustive search. Instead, we applied a stochastic tree-searching algorithm.  

Q3.4 Review the Google Slides from Week 2. What is the total number of trees that exist in the tree universe for the MSA in this assignment? (3 points)  

